pkg/frontend/transport/handler.go

```diff
@@ -134,19 +134,41 @@ func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		_ = r.Body.Close()
 	}()
 
-	// Buffer the body for later use to track slow queries.
-	var buf bytes.Buffer
-	r.Body = http.MaxBytesReader(w, r.Body, f.cfg.MaxBodySize)
-	r.Body = io.NopCloser(io.TeeReader(r.Body, &buf))
+	// Store the body contents in a seeker, so we can read it multiple times.
+	seeker, err := readIntoReadCloseSeeker(http.MaxBytesReader(w, r.Body, f.cfg.MaxBodySize))
+	if err != nil {
+		writeError(w, err)
+		return
+	}
+	r.Body = seeker
+
+	// Parse the form, as it's needed to build the activity for the activity-tracker.
+	if err := r.ParseForm(); err != nil {
+		writeError(w, apierror.New(apierror.TypeBadData, err.Error()))
+		return
+	}
+
+	// Store the r.Form contents copy, as middlewares may modify it (like setting the default values).
+	params := copyValues(r.Form)
+
+	// Seek the body back to the beginning, so it can be read again if it's used to forward the request through a roundtripper.
```

Comment:
I can't see why it's a red flag. 

> it leaves Request in the state which handler or middlewares may not expect

Neither handler or middlewares should expect undocumented states. There's no docs saying that a non-nil `r.Form` indicates an empty body.

I don't think that we need to parse twice just because someone takes a wrong assumption somewhere else.

Reply:
> There's no docs saying that a non-nil r.Form indicates an empty body.

non-nil `r.Form` implies that `ParseForm` was called (based on `Form` field documentation), and `ParseForm` is documented to read the body for POST, PUT or PATCH request with `Content-Type: application/x-www-form-urlencoded`.

Since our middlewares don't currently parse the form, we wouldn't parse it twice.

Anyway, well behaved middlewares (ie. all in Mimir) are not making any such assumptions, so we don't need to worry about it anymore.